crates/nu-protocol/src/engine/engine_state.rs

```diff
@@ -1551,6 +1551,22 @@ impl<'a> StateWorkingSet<'a> {
         num_permanent_vars + self.delta.vars.len()
     }
 
+    pub fn list_variables(&self) -> Vec<&[u8]> {
```

Comment:
This is not a correct way to list variables for two reasons:
1) If a variable is shadowed, it will be listed twice, you might want to check for duplicates before pushing to the vector, or use HashSet (in that case, iterating forward, instead of backward would be better), or something similar.
2) You're using only variables in the working set, you should search through the ones in the permanent state (see `find_variable()` below)

Reply:
thanks! I'll try those. I think HashSet makes sense here given that all we care about is presence or absence of variables.

A thing I was playing around with was whether there was a way to create an iterator over all the valid variables rather than a vector. The thought was that iterator could then be reused in a few other places in `engine_state` (including `find_variable`). I was having some trouble getting the lifetimes to line up in a sane way, but does that sound like something that would be useful or just overkill?

Reply:
It seems like an overkill for now. We're looking to re-implement the parser and the scoping/overlay logic so it's not worth investing too much time on it, it's likely going to be rewritten.